src/storage-types/src/sources.rs

```diff
@@ -838,6 +842,130 @@ pub enum GenericSourceConnection<C: ConnectionAccess = InlinedConnection> {
+                    id,
+                    external_frontier,
+                    remap_subscribe,
+                )
+                .await
+            }
+            GenericSourceConnection::Postgres(pg) => {
+                let external_frontier = pg
+                    .fetch_write_frontier(&config)
+                    .await
+                    .map_err(StorageError::Generic)?;
+
+                GenericSourceConnection::decode_remap_data_until_geq_external_frontier(
+                    id,
+                    external_frontier,
+                    remap_subscribe,
+                )
+                .await
+            }
+            GenericSourceConnection::MySql(my_sql) => {
+                let external_frontier = my_sql
+                    .fetch_write_frontier(&config)
+                    .await
+                    .map_err(StorageError::Generic)?;
+
+                GenericSourceConnection::decode_remap_data_until_geq_external_frontier(
+                    id,
+                    external_frontier,
+                    remap_subscribe,
+                )
+                .await
+            }
+            s @ GenericSourceConnection::LoadGenerator(_) => unreachable!(
+                "do not try to determine RTR timestamp on {} source",
+                s.name()
+            ),
+        }
+    }
+
+    async fn decode_remap_data_until_geq_external_frontier<
+        ExternalFrontier: SourceTimestamp,
+        T: Timestamp + Lattice + Codec64 + From<EpochMillis>,
+    >(
+        id: GlobalId,
+        external_frontier: timely::progress::Antichain<ExternalFrontier>,
+        mut remap_subscribe: Subscribe<SourceData, (), T, mz_repr::Diff>,
+    ) -> Result<T, StorageError<T>> {
+        use mz_persist_client::read::ListenEvent;
+        use mz_timely_util::antichain::AntichainExt;
+        use timely::progress::frontier::MutableAntichain;
+
+        tracing::debug!(
+            ?id,
+            "fetched real time recency frontier: {}",
+            external_frontier.pretty()
+        );
+
+        let mut native_upper = MutableAntichain::new();
+        let mut remap_updates = vec![];
+
+        loop {
+            for event in remap_subscribe.fetch_next().await {
+                match event {
+                    ListenEvent::Updates(updates) => {
+                        for ((k, v), t, d) in updates {
+                            let row: Row = k.expect("invalid binding").0.expect("invalid binding");
+                            let _v: () = v.expect("invalid binding");
+                            let from_ts: ExternalFrontier = SourceTimestamp::decode_row(&row);
+                            remap_updates.push((from_ts, t, d));
+                        }
+                    }
+                    ListenEvent::Progress(f) => {
+                        let updates = Vec::with_capacity(remap_updates.len());
+                        let (max_ts, updates) = remap_updates.drain(..).fold(
```

Comment:
There are a number of issues with this code. Here we are processing all updates in `remap_updates` even if they are beyond the frontier `f`. Also, we are only considering the maximum timestamp observed in the batch which might produce a non-tight timestamp. Here is a diff with a way to do it that guards against these edge cases:

```diff
diff --git a/src/storage-types/src/sources.rs b/src/storage-types/src/sources.rs
index 222091e7a1..beccac61bf 100644
--- a/src/storage-types/src/sources.rs
+++ b/src/storage-types/src/sources.rs
@@ -9,7 +9,9 @@
 
 //! Types and traits related to the introduction of changing collections into `dataflow`.
 
-use std::collections::BTreeMap;
+use std::cmp::Reverse;
+use std::collections::binary_heap::PeekMut;
+use std::collections::{BTreeMap, BinaryHeap};
 use std::fmt::Debug;
 use std::hash::Hash;
 use std::ops::{Add, AddAssign, Deref, DerefMut};
@@ -44,7 +46,7 @@ use prost::Message;
 use serde::{Deserialize, Serialize};
 use timely::order::{PartialOrder, TotalOrder};
 use timely::progress::timestamp::Refines;
-use timely::progress::{PathSummary, Timestamp};
+use timely::progress::{Antichain, PathSummary, Timestamp};
 
 use crate::configuration::StorageConfiguration;
 use crate::connections::inline::{
@@ -847,6 +849,7 @@ impl GenericSourceConnection {
         self,
         id: GlobalId,
         config: StorageConfiguration,
+        as_of: Antichain<T>,
         remap_subscribe: Subscribe<SourceData, (), T, mz_repr::Diff>,
     ) -> Result<T, StorageError<T>> {
         match self {
@@ -859,6 +862,7 @@ impl GenericSourceConnection {
                 GenericSourceConnection::decode_remap_data_until_geq_external_frontier(
                     id,
                     external_frontier,
+                    as_of,
                     remap_subscribe,
                 )
                 .await
@@ -872,6 +876,7 @@ impl GenericSourceConnection {
                 GenericSourceConnection::decode_remap_data_until_geq_external_frontier(
                     id,
                     external_frontier,
+                    as_of,
                     remap_subscribe,
                 )
                 .await
@@ -885,6 +890,7 @@ impl GenericSourceConnection {
                 GenericSourceConnection::decode_remap_data_until_geq_external_frontier(
                     id,
                     external_frontier,
+                    as_of,
                     remap_subscribe,
                 )
                 .await
@@ -897,11 +903,12 @@ impl GenericSourceConnection {
     }
 
     async fn decode_remap_data_until_geq_external_frontier<
-        ExternalFrontier: SourceTimestamp,
-        T: Timestamp + Lattice + Codec64 + From<EpochMillis>,
+        FromTime: SourceTimestamp,
+        T: Timestamp + Lattice + Codec64,
     >(
         id: GlobalId,
-        external_frontier: timely::progress::Antichain<ExternalFrontier>,
+        external_frontier: timely::progress::Antichain<FromTime>,
+        as_of: Antichain<T>,
         mut remap_subscribe: Subscribe<SourceData, (), T, mz_repr::Diff>,
     ) -> Result<T, StorageError<T>> {
         use mz_persist_client::read::ListenEvent;
@@ -914,55 +921,76 @@ impl GenericSourceConnection {
             external_frontier.pretty()
         );
 
+        let external_frontier = external_frontier.borrow();
         let mut native_upper = MutableAntichain::new();
-        let mut remap_updates = vec![];
 
-        loop {
+        let mut remap_frontier = Antichain::from_elem(T::minimum());
+        let mut pending_remap = BinaryHeap::new();
+
+        let mut min_ts = T::minimum();
+        min_ts.advance_by(as_of.borrow());
+        let mut min_ts = Some(min_ts);
+
+        // First we must read the snapshot so that we can accumulate the minimum timestamp
+        // correctly.
+        while !remap_frontier.is_empty() {
+            // Receive binding updates
             for event in remap_subscribe.fetch_next().await {
                 match event {
                     ListenEvent::Updates(updates) => {
-                        for ((k, v), t, d) in updates {
+                        for ((k, v), into_ts, diff) in updates {
                             let row: Row = k.expect("invalid binding").0.expect("invalid binding");
                             let _v: () = v.expect("invalid binding");
-                            let from_ts: ExternalFrontier = SourceTimestamp::decode_row(&row);
-                            remap_updates.push((from_ts, t, d));
+                            let from_ts: FromTime = SourceTimestamp::decode_row(&row);
+                            pending_remap.push(Reverse((into_ts, from_ts, diff)));
                         }
                     }
-                    ListenEvent::Progress(f) => {
-                        let updates = Vec::with_capacity(remap_updates.len());
-                        let (max_ts, updates) = remap_updates.drain(..).fold(
-                            (T::minimum(), updates),
-                            |(ts, mut updates), (from_ts, t, d)| {
-                                updates.push((from_ts, d));
-                                (std::cmp::max(ts, t), updates)
-                            },
-                        );
-
-                        // If there were no updates, then the ts we get is not
-                        // derived from any remap data.
-                        if updates.is_empty() && !f.is_empty() {
-                            continue;
+                    ListenEvent::Progress(frontier) => remap_frontier = frontier,
+                }
+            }
+            // Only process the data if we can correctly accumulate timestamps beyond the as_of
+            if as_of.iter().any(|t| !remap_frontier.less_equal(t)) {
+                loop {
+                    // The next timestamp that must be considered is either the minimum timestamp,
+                    // if we haven't yet checked it, or the next timestamp of the pending remap
+                    // updates that is not beyond the frontier.
+                    let binding_ts = match min_ts.take() {
+                        Some(min_ts) => min_ts,
+                        None => {
+                            let Some(Reverse((into_ts, _, _))) = pending_remap.peek() else {
+                                break;
+                            };
+                            if !remap_frontier.less_equal(into_ts) {
+                                into_ts.clone()
+                            } else {
+                                break;
+                            }
                         }
-
-                        native_upper.update_iter(updates);
-
-                        // At this TS, the source contained all of the data visible
-                        // in the external system.
-                        if PartialOrder::less_equal(
-                            &external_frontier.borrow(),
-                            &native_upper.frontier(),
-                        ) {
-                            tracing::trace!("real-time recency ts for {id}: {max_ts:?}");
-                            return Ok(max_ts);
-                        } else if f.is_empty() {
-                            // Collection dropped before we ingested external
-                            // frontier.
-                            return Err(StorageError::RtrDropFailure(id));
+                    };
+                    // Create an iterator with all the updates that are less than or equal to
+                    // binding_ts
+                    let binding_updates = std::iter::from_fn(|| {
+                        let update = pending_remap.peek_mut()?;
+                        if PartialOrder::less_equal(&update.0 .0, &binding_ts) {
+                            let Reverse((_, from_ts, diff)) = PeekMut::pop(update);
+                            Some((from_ts, diff))
+                        } else {
+                            None
                         }
+                    });
+                    native_upper.update_iter(binding_updates);
+
+                    // Check if at binding_ts the source contained all of the data visible in the
+                    // external system. If so, we're done.
+                    if PartialOrder::less_equal(&external_frontier, &native_upper.frontier()) {
+                        tracing::trace!("real-time recency ts for {id}: {binding_ts:?}");
+                        return Ok(binding_ts);
                     }
                 }
             }
         }
+        // Collection dropped before we ingested external frontier.
+        Err(StorageError::RtrDropFailure(id))
     }
 }
 
```

Reply:
Thanks for this; couldn't have come up with this implementation myself.